File: hate_system.py
"""
Project Elements - Hate & Target System (Phase 3)

実効ヘイト計算とターゲット選定ロジック。
後衛全滅時は生存者の最後尾ポジションを 2.0x に動的繰り上げする。
"""
from __future__ import annotations
from typing import List, Optional

from models import Character, Party, Position

# ポジションの「後ろ度」順序 (大きいほど後ろ)
POSITION_ORDER: dict[Position, int] = {
    Position.FRONT: 0,
    Position.MID: 1,
    Position.BACK: 2,
}


class HateSystem:
    """
    ヘイト管理とターゲット選定を担う静的ユーティリティ。

    動的ポジション補正の考え方:
      「生存者の中で最後尾にいるポジション」は常に 2.0x を得る。
      後衛が全員死んでいれば中衛が 2.0x に繰り上がり、
      中衛まで全滅すれば前衛が 2.0x になる。
    """

    @staticmethod
    def get_dynamic_multiplier(
        character: Character,
        alive_chars: List[Character],
    ) -> float:
        """
        生存者リストを基に動的なポジション補正倍率を返す。

        - 生存者の中で最後尾のポジション → 2.0x（本来の倍率に関わらず）
        - それ以外 → Position に定義された固定倍率
        """
        if not alive_chars:
            return character.position.hate_multiplier

        backmost_order = max(POSITION_ORDER[c.position] for c in alive_chars)
        if POSITION_ORDER[character.position] == backmost_order:
            return 2.0
        return character.position.hate_multiplier

    @classmethod
    def get_effective_hate(
        cls,
        character: Character,
        alive_chars: List[Character],
    ) -> float:
        """実効ヘイトスコア = 基礎ヘイト値 × 動的ポジション補正倍率"""
        return character.base_hate * cls.get_dynamic_multiplier(character, alive_chars)

    @classmethod
    def select_target(cls, characters: List[Character]) -> Optional[Character]:
        """
        生存キャラクターの中から攻撃ターゲットを選択する。

        優先順位:
          1. 実効ヘイトスコアが最大
          2. 同値ならポジションが後ろ（POSITION_ORDER 降順）
          3. さらに同値なら残りHP割合が低い方
        """
        alive = [c for c in characters if c.is_alive]
        if not alive:
            return None

        def sort_key(c: Character):
            hate = cls.get_effective_hate(c, alive)
            return (-hate, -POSITION_ORDER[c.position], c.hp_ratio)

        return min(alive, key=sort_key)

    @staticmethod
    def add_hate(character: Character, amount: float) -> None:
        """行動によるヘイト蓄積（基礎ヘイト値を加算する）。"""
        character.base_hate += amount

File: Ver-C#/Scripts/Logic/HateSystem.cs
// Ver-C#/Scripts/Logic/HateSystem.cs
// hate_system.py の移植。
// 動的なポジション補正倍率の算出、実効ヘイトスコアの計算、
// およびタイブレークロジックを含むターゲット選定を担う。

using System.Collections.Generic;
using System.Linq;
using ProjectElements.Core;
using ProjectElements.Models;

namespace ProjectElements.Logic
{
    /// <summary>
    /// ヘイト管理とターゲット選定を担う静的ユーティリティクラス。
    /// Python の HateSystem クラスに相当。
    ///
    /// 動的ポジション補正の考え方:
    ///   「生存者の中で最後尾にいるポジション」は常に 2.0x を得る。
    ///   後衛が全員戦闘不能なら中衛が 2.0x に繰り上がり、
    ///   中衛まで全滅すれば前衛が 2.0x になる。
    /// </summary>
    public static class HateSystem
    {
        // ---------------------------------------------------------
        // ① 動的ポジション補正倍率の取得
        // ---------------------------------------------------------

        /// <summary>
        /// 生存者リストを基に、そのキャラクターの動的なポジション補正倍率を返す。
        ///
        /// ・生存者の中で最後尾ポジション（GetOrderIndex が最大）のキャラ → 2.0x
        /// ・それ以外 → Position に定義された固定倍率（Front=0.5, Mid=1.0, Back=2.0）
        ///
        /// Python の get_dynamic_multiplier() に相当。
        /// </summary>
        public static float GetDynamicMultiplier(
            BattleCharacter              character,
            IReadOnlyList<BattleCharacter> aliveChars)
        {
            // 生存者がいない場合は固定倍率をそのまま返す（安全策）
            if (aliveChars.Count == 0)
                return character.Position.GetBaseHateMultiplier();

            // 生存者の中で最も後方のポジション順序を取得
            int backmostOrder = aliveChars.Max(c => c.Position.GetOrderIndex());

            // このキャラクターが最後尾なら 2.0x に繰り上げ
            if (character.Position.GetOrderIndex() == backmostOrder)
                return 2.0f;

            // それ以外は固定倍率
            return character.Position.GetBaseHateMultiplier();
        }

        // ---------------------------------------------------------
        // ② 実効ヘイトスコアの算出
        // ---------------------------------------------------------

        /// <summary>
        /// 実効ヘイトスコア = 基礎ヘイト値 × 動的ポジション補正倍率。
        /// ターゲット選定の主比較値として使用する。
        ///
        /// Python の get_effective_hate() に相当。
        /// </summary>
        public static float GetEffectiveHate(
            BattleCharacter              character,
            IReadOnlyList<BattleCharacter> aliveChars)
        {
            return character.BaseHate * GetDynamicMultiplier(character, aliveChars);
        }

        // ---------------------------------------------------------
        // ③ ターゲット選定
        // ---------------------------------------------------------

        /// <summary>
        /// 生存キャラクターの中から敵の攻撃ターゲットを選択して返す。
        /// 生存者がいない場合は null を返す。
        ///
        /// タイブレーク優先順位:
        ///   1. 実効ヘイトスコアが最大
        ///   2. 同値ならポジションが後方（GetOrderIndex が大きい）
        ///   3. さらに同値なら残りHP割合が低い方
        ///
        /// Python の select_target() に相当。
        /// </summary>
        public static BattleCharacter SelectTarget(IReadOnlyList<BattleCharacter> characters)
        {
            var alive = characters.Where(c => c.IsAlive).ToList();
            if (alive.Count == 0) return null;

            // LINQ の OrderBy チェーンでタイブレークを再現
            // Python の sort_key: (-hate, -position_order, hp_ratio) に対応
            return alive
                .OrderByDescending(c => GetEffectiveHate(c, alive))   // 1. ヘイト降順
                .ThenByDescending(c => c.Position.GetOrderIndex())      // 2. ポジション後方優先
                .ThenBy          (c => c.HpRatio)                       // 3. HP割合昇順（低い方を優先）
                .First();
        }

        // ---------------------------------------------------------
        // ④ ヘイト蓄積
        // ---------------------------------------------------------

        /// <summary>
        /// キャラクターが行動するたびに呼び出し、基礎ヘイト値を加算する。
        /// Python の add_hate() に相当。
        /// </summary>
        public static void AddHate(BattleCharacter character, float amount)
        {
            character.BaseHate += amount;
        }
    }
}
